File: isolation/contract/runtime.go
package contract

import (
	"crypto/sha256"
	"errors"
	"fmt"
)

// The runtime that compiles and runs the artifact INSIDE the domain (decided 2026-09-23): the app is
// distributed as a portable WebAssembly component and JIT-compiled to the domain's own ISA after the
// bundle is verified. What that runtime is, and how it is configured, is part of what a report vouches
// for: two domains running the same bundle on different runtimes, ISAs or CPU-feature policies are
// different things to a verifier.
//
// Requirements this type encodes (fail closed: an identity that cannot state them is refused):
//   - the JIT and runtime execute inside the protected, measured boundary;
//   - W^X: no page is ever both writable and executable (WX must say "enforced");
//   - no host-supplied native code and no unverified compiled cache: Cache is "none" (compile every
//     time) or "authenticated" (a cache keyed by CacheKey and authenticated under a domain-held key);
//   - the runtime name, version, target ISA and CPU-feature policy are bound into the report via Bind2.
//
// Execution has two admissible modes, because a stock Pixel protected VM refuses every executable
// mapping (measured 2026-09-23 on a Pixel 10 Pro XL: execmem denied, RWX and RW->RX both refused, no
// writable memfd, the data store noexec): there the runtime compiles the verified component with
// Cranelift to wasmtime's Pulley bytecode, still inside the boundary, and INTERPRETS it. W^X then holds
// trivially. Everywhere else the runtime JITs to the host's own ISA. The mode is part of the identity:
//   - "jit":         TargetISA is the host's ISA (x86_64 | aarch64) and equals HostISA;
//   - "interpreter": TargetISA is "pulley64" and HostISA names the real hardware the interpreter runs on,
//                    so the CPU-feature policy still describes something a verifier can reason about.
type RuntimeIdentity struct {
	Name        string `json:"name"`        // e.g. "wasmtime"
	Version     string `json:"version"`     // e.g. "48.0.1"
	Execution   string `json:"execution"`   // "jit" | "interpreter"
	TargetISA   string `json:"targetIsa"`   // "x86_64" | "aarch64" (jit) | "pulley64" (interpreter)
	HostISA     string `json:"hostIsa"`     // "x86_64" | "aarch64": the ISA the runtime itself executes on
	CPUFeatures string `json:"cpuFeatures"` // the enabled feature policy, a canonical string ("baseline", or "+sse4.2,+avx2", ...)
	WX          string `json:"wx"`          // "enforced"
	Cache       string `json:"cache"`       // "none" | "authenticated"
}

const (
	ABI2 = "enclave-domain-abi/2" // ABI with the runtime identity bound into the report

	ISAx86_64   = "x86_64"
	ISAaarch64  = "aarch64"
	ISApulley64 = "pulley64" // wasmtime's portable bytecode target: interpreter only

	ExecJIT         = "jit"
	ExecInterpreter = "interpreter"

	WXEnforced         = "enforced"
	CacheNone          = "none"
	CacheAuthenticated = "authenticated"
)

var bind2Domain = []byte("enclave-bind-v2\n")
var cacheKeyDomain = []byte("enclave-compiled-cache-v1\n")

// Validate refuses an identity that does not meet the requirements above.
func (r RuntimeIdentity) Validate() error {
	if r.Name == "" || r.Version == "" {
		return errors.New("runtime name and version are required")
	}
	if r.HostISA != ISAx86_64 && r.HostISA != ISAaarch64 {
		return fmt.Errorf("host ISA %q is not one of %s, %s", r.HostISA, ISAx86_64, ISAaarch64)
	}
	switch r.Execution {
	case ExecJIT:
		if r.TargetISA != r.HostISA {
			return fmt.Errorf("a JIT emits the host's own ISA: target %q must equal host %q", r.TargetISA, r.HostISA)
		}
	case ExecInterpreter:
		if r.TargetISA != ISApulley64 {
			return fmt.Errorf("an interpreter runs %s bytecode, not %q", ISApulley64, r.TargetISA)
		}
	default:
		return fmt.Errorf("execution %q is not one of %s, %s", r.Execution, ExecJIT, ExecInterpreter)
	}
	if r.CPUFeatures == "" {
		return errors.New("the CPU-feature policy must be stated (\"baseline\" if none)")
	}
	if r.WX != WXEnforced {
		return errors.New("a runtime that cannot state W^X as enforced is not admissible")
	}
	if r.Cache != CacheNone && r.Cache != CacheAuthenticated {
		return fmt.Errorf("cache mode %q is not one of %s, %s", r.Cache, CacheNone, CacheAuthenticated)
	}
	return nil
}

// RuntimeID is the identity's digest: sha256 of its canonical JSON. Same fields, same ID, whatever
// encoder wrote them.
func RuntimeID(r RuntimeIdentity) ([32]byte, error) {
	if err := r.Validate(); err != nil {
		return [32]byte{}, err
	}
	b, err := Canonical(r)
	if err != nil {
		return [32]byte{}, err
	}
	return sha256.Sum256(b), nil
}

// Bind2 is the ABI/2 binding: the domain's key, the verifier's nonce AND the runtime identity, in one
// 32-byte value for report_data[0:32]. report_data[32:64] stays the app ID. A verifier recomputes it
// from the SPKI its own handshake saw, its own nonce, and the runtime identity the domain states in its
// attestation document; a document naming a different runtime, version, ISA or feature policy than the
// one that asked for the report does not verify.
func Bind2(spki, nonce []byte, runtimeID [32]byte) ([32]byte, error) {
	if len(nonce) != NonceLen {
		return [32]byte{}, ErrNonceLen
	}
	h := sha256.New()
	h.Write(bind2Domain)
	h.Write(spki)
	h.Write(nonce)
	h.Write(runtimeID[:])
	var out [32]byte
	copy(out[:], h.Sum(nil))
	return out, nil
}

// CacheKey names a compiled artifact a domain may keep: the bundle's identity and the runtime identity
// (which includes the target ISA and the CPU-feature policy) together. A cache entry under any other key,
// or one that fails its authentication, is rebuilt inside the domain rather than used.
func CacheKey(appID, runtimeID [32]byte) [32]byte {
	h := sha256.New()
	h.Write(cacheKeyDomain)
	h.Write(appID[:])
	h.Write(runtimeID[:])
	var out [32]byte
	copy(out[:], h.Sum(nil))
	return out
}

File: isolation/contract/runtime.mjs
// The verifier's half of runtime.go: the runtime identity, its digest, the ABI/2 binding and the
// compiled-cache key, in JavaScript.
//
// Why a mirror and not a re-implementation: a verifier has to recompute report_data[0:32] from the
// identity a domain STATES in its attestation document. If the two sides disagree by a byte - a field
// order, a missing field, a laxer validation - every honest domain looks like a liar, or worse, a
// domain claiming an inadmissible runtime passes. So this file is driven by isolation/contract/
// vectors.json, the same vectors the Go implementation and the Rust launcher pass
// (test/isolation-runtime-identity.test.mjs), and the fail-closed rules are copied from Validate()
// deliberately rather than rephrased.
//
// The rules, from isolation/contract/RUNTIME.md:
//   - execution "jit": the target ISA is the host's own (x86_64 | aarch64) and equals hostIsa;
//   - execution "interpreter": the target is pulley64 and hostIsa names the real hardware. This is the
//     mode a domain that may not hold executable pages uses - a stock Pixel pVM (execmem denied,
//     measured) and a Windows VBS enclave (ERROR_DYNAMIC_CODE_BLOCKED, measured in windows/PARITY.md);
//   - W^X must be stated as enforced, the cache must be none or authenticated, and the CPU-feature
//     policy must be stated. An identity that cannot say these is refused, not downgraded.
import { createHash } from 'node:crypto';

export const ABI1 = 'enclave-domain-abi/1';
export const ABI2 = 'enclave-domain-abi/2';

export const ISA_X86_64 = 'x86_64';
export const ISA_AARCH64 = 'aarch64';
export const ISA_PULLEY64 = 'pulley64';
export const EXEC_JIT = 'jit';
export const EXEC_INTERPRETER = 'interpreter';
export const WX_ENFORCED = 'enforced';
export const CACHE_NONE = 'none';
export const CACHE_AUTHENTICATED = 'authenticated';

const BIND2_DOMAIN = Buffer.from('enclave-bind-v2\n');
const CACHE_KEY_DOMAIN = Buffer.from('enclave-compiled-cache-v1\n');

// the fields of RuntimeIdentity, and nothing else: an unknown field is a different identity than the
// one the Go type describes, so it is refused rather than ignored
const FIELDS = ['name', 'version', 'execution', 'targetIsa', 'hostIsa', 'cpuFeatures', 'wx', 'cache'];

// canonical is Canonical() from bundle.go: compact JSON, keys sorted at every level, no HTML escaping.
export function canonical(v) {
  const sort = (x) => {
    if (Array.isArray(x)) return x.map(sort);
    if (x && typeof x === 'object') {
      const out = {};
      for (const k of Object.keys(x).sort()) out[k] = sort(x[k]);
      return out;
    }
    return x;
  };
  return Buffer.from(JSON.stringify(sort(v)));
}

// validateRuntimeIdentity returns null when the identity is admissible, or the reason it is not.
// Fail closed: every reason is a refusal, never a downgrade.
export function validateRuntimeIdentity(r) {
  if (!r || typeof r !== 'object' || Array.isArray(r)) return 'the runtime identity is not an object';
  for (const k of Object.keys(r)) {
    if (!FIELDS.includes(k)) return `the runtime identity carries an unknown field ${JSON.stringify(k)}`;
  }
  for (const k of FIELDS) {
    if (typeof r[k] !== 'string') return `the runtime identity field ${k} is missing or not a string`;
  }
  if (r.name === '' || r.version === '') return 'runtime name and version are required';
  if (r.hostISA === '') return 'the host ISA must be stated';
  if (r.hostIsa !== ISA_X86_64 && r.hostIsa !== ISA_AARCH64) {
    return `host ISA ${JSON.stringify(r.hostIsa)} is not one of ${ISA_X86_64}, ${ISA_AARCH64}`;
  }
  if (r.execution === EXEC_JIT) {
    if (r.targetIsa !== r.hostIsa) {
      return `a JIT emits the host's own ISA: target ${JSON.stringify(r.targetIsa)} must equal host ${JSON.stringify(r.hostIsa)}`;
    }
  } else if (r.execution === EXEC_INTERPRETER) {
    if (r.targetIsa !== ISA_PULLEY64) {
      return `an interpreter runs ${ISA_PULLEY64} bytecode, not ${JSON.stringify(r.targetIsa)}`;
    }
  } else {
    return `execution ${JSON.stringify(r.execution)} is not one of ${EXEC_JIT}, ${EXEC_INTERPRETER}`;
  }
  if (r.cpuFeatures === '') return 'the CPU-feature policy must be stated ("baseline" if none)';
  if (r.wx !== WX_ENFORCED) return 'a runtime that cannot state W^X as enforced is not admissible';
  if (r.cache !== CACHE_NONE && r.cache !== CACHE_AUTHENTICATED) {
    return `cache mode ${JSON.stringify(r.cache)} is not one of ${CACHE_NONE}, ${CACHE_AUTHENTICATED}`;
  }
  return null;
}

// runtimeId is sha256 of the identity's canonical JSON, exactly RuntimeID() in runtime.go. An
// inadmissible identity has no ID: it throws rather than returning a digest something could pin.
export function runtimeId(r) {
  const why = validateRuntimeIdentity(r);
  if (why) throw new Error(`inadmissible runtime identity: ${why}`);
  const only = {};
  for (const k of FIELDS) only[k] = r[k];
  return createHash('sha256').update(canonical(only)).digest();
}

// bind2 is the ABI/2 binding for report_data[0:32]: the domain's transport key, the verifier's nonce
// and the runtime identity together. report_data[32:64] stays the app ID.
export function bind2(spki, nonce, rid) {
  if (nonce.length !== 32) throw new Error('nonce must be 32 bytes');
  if (rid.length !== 32) throw new Error('runtime id must be 32 bytes');
  return createHash('sha256').update(Buffer.concat([BIND2_DOMAIN, Buffer.from(spki), Buffer.from(nonce), Buffer.from(rid)])).digest();
}

// bind1 is the ABI/1 binding, kept here so a caller can compute either from one place.
export function bind1(spki, nonce) {
  if (nonce.length !== 32) throw new Error('nonce must be 32 bytes');
  return createHash('sha256').update(Buffer.concat([Buffer.from(spki), Buffer.from(nonce)])).digest();
}

// cacheKey names a compiled artifact a domain may keep: bundle identity and runtime identity together,
// so the same component compiled by another runtime version, for another ISA or under another feature
// policy is a different entry. Anything under another key, or failing its authentication, is rebuilt.
export function cacheKey(appId, rid) {
  return createHash('sha256').update(Buffer.concat([CACHE_KEY_DOMAIN, Buffer.from(appId), Buffer.from(rid)])).digest();
}
